src/shared/elf-util.c

```diff
@@ -0,0 +1,812 @@
+                                 * JSON blob. */
+                                r = json_build(&w, JSON_BUILD_OBJECT(JSON_BUILD_PAIR(name, JSON_BUILD_VARIANT(v))));
+                                if (r < 0)
+                                        return log_error_errno(r, "Failed to build JSON object: %m");
+                                r = json_variant_merge(c->package_metadata, w);
+                                if (r < 0)
+                                        return log_error_errno(r, "json_variant_merge of package meta with buildid failed: %m");
+
+                                /* Finally stash the name, so we avoid double visits. */
+                                r = set_put_strdup(c->modules, name);
+                                if (r < 0)
+                                        return log_error_errno(r, "set_put_strdup failed: %m");
+
+                                return 1;
+                        }
+                }
+        }
+
+        /* Didn't find package metadata for this module - that's ok, just go to the next. */
+        return 0;
+}
+
+/* Get the build-id out of an ELF object or a dwarf core module. */
+static int parse_buildid(Dwfl_Module *mod, Elf *elf, const char *name, struct stack_context *c, JsonVariant **ret_id_json) {
+        _cleanup_(json_variant_unrefp) JsonVariant *id_json = NULL;
+        const unsigned char *id;
+        GElf_Addr id_vaddr;
+        ssize_t id_len;
+        int r;
+
+        assert(mod || elf);
+        assert(c);
+
+        if (mod)
+                id_len = sym_dwfl_module_build_id(mod, &id, &id_vaddr);
+        else
+                id_len = sym_dwelf_elf_gnu_build_id(elf, (const void **)&id);
+        if (id_len <= 0) {
+                /* If we don't find a build-id, note it in the journal message, and try
+                 * anyway to find the package metadata. It's unlikely to have the latter
+                 * without the former, but there's no hard rule. */
+                if (c->f)
+                        fprintf(c->f, "Found module %s without build-id.\n", name);
+        } else {
+                /* We will later parse package metadata json and pass it to our caller. Prepare the
+                * build-id in json format too, so that it can be appended and parsed cleanly. It
+                * will then be added as metadata to the journal message with the stack trace. */
+                r = json_build(&id_json, JSON_BUILD_OBJECT(JSON_BUILD_PAIR("buildId", JSON_BUILD_HEX(id, id_len))));
+                if (r < 0)
+                        return log_error_errno(r, "json_build on build-id failed: %m");
+
+                if (c->f) {
+                        JsonVariant *build_id = json_variant_by_key(id_json, "buildId");
+                        assert_se(build_id);
+                        fprintf(c->f, "Found module %s with build-id: %s\n", name, json_variant_string(build_id));
```

Comment:
Maybe `"Module %s with build-id %s."`  ("Found" just uses up space, and ":" is not necessary.")